### ISSUE TITLE
Question About Larger LoRA Rank Causing Some MLP LoRA Weights to Be Empty

### ISSUE THREAD

**Issue by xinyann666** [franciscoliu/MLLMU-Bench#6]
I’m sorry to trouble you again. I am still working on reproducing your repo and encountered an issue when setting LoRA rank (r) to 16 or 32. Specifically, I noticed that the LoRA checkpoint contains empty tensors (torch.Size([0])) for the following parameters across all layers in the LLM backbone for LLaVA:
mlp.up_proj.lora_B.weight
mlp.gate_proj.lora_B.weight
mlp.down_proj.lora_A.weight
However, this issue is resolved when I change the rank to 8 in both finetune.py and the baselines. I’m not sure why this happens and it is so strange.
I am reaching out for any advice on this and also want to check whether the LoRA rank setting is important here. I noticed that in finetune.py, r=8, while in all baselines, r=16. 
I hope I’m not being annoying—I truly appreciate your help!

**Comment by franciscoliu** [comment]
Thanks for reaching out, and no worries at all—I appreciate you taking the time to reproduce the repo!

Regarding the issue with LoRA rank (r), it does seem unusual that the tensors are empty for ranks 16 and 32 but work fine for rank 8. A few things you might check:

- Checkpoint Initialization: Ensure that the checkpoint is correctly initialized for different LoRA ranks. If the model is loading an incompatible checkpoint (e.g., one trained with a different rank), it could result in empty tensors.

- LoRA Implementation Differences: The behavior might depend on how LLaVA integrates LoRA. Some implementations dynamically allocate LoRA weights based on rank, and if there's a misalignment in initialization or weight loading, it could cause issues for certain values.

- Memory Constraints: Higher ranks require more memory. If there are constraints during training or checkpoint saving, the tensors might not be correctly allocated. You could try running with torch.cuda.empty_cache() before training and monitoring memory usage.

- Config Mismatch: You mentioned that r=8 is set in finetune.py, while the baselines use r=16. Double-check that all configurations are consistent when switching ranks. If the model was originally trained with r=8 and you're now switching to r=16 or r=32, try reinitializing and training from scratch rather than loading a checkpoint trained with a different rank.

- As for whether LoRA rank is important—yes, it does impact the expressiveness of the adaptation. Higher ranks generally allow for more capacity in low-rank adaptation but also require more memory. In practice, r=8 is often sufficient, but r=16 may provide better fine-tuning performance depending on the task.
- Lastly, I recommend checking the resolved issues in this repo, as I recall that someone has successfully reproduced the results before. You may want to check if there are any configuration mismatches or missing dependencies causing the issue.

If you're still facing this issue, **could you try training with r=16 from scratch (without loading an existing checkpoint)** and see if the tensors are correctly populated? Let me know what you find, and I’d be happy to help further!

**Comment by xinyann666** [comment]
Thank you so much! 
I still get empty tensors when I set the LoRA rank higher than 8, even when fine-tuning the original `llava-v1.5-hf` checkpoint. I suspect this might be due to my GPU memory constraints. However, while trying to solve this issue, I noticed that the current code also fine-tunes the vision encoder. When I print all parameter names in the adapter checkpoint, I see that the vision tower is not frozen.
I noticed that `find_all_linear_names(model)` comes from the original LLaVA repo, but I am quite confused about its behavior. I also found a related issue in the original repo: [Issue #1647](https://github.com/haotian-liu/LLaVA/issues/1647). Since this function only stores the last keywords:
['gate_proj', 'o_proj', 'up_proj', 'v_proj', 'down_proj', 'k_proj', 'q_proj']
when passed into lora_config, some modules of the vision encoder seem to match as well—although these modules were originally meant to be excluded for LoRA.

```python
My code:
sd = load_file("./model/test-LoRA/adapter_model.safetensors")
for key, val in sd.items():
    print(key, val.shape)
Part of the output:
base_model.model.vision_tower.vision_model.encoder.layers.9.self_attn.k_proj.lora_A.weight torch.Size([8, 1024])
base_model.model.vision_tower.vision_model.encoder.layers.9.self_attn.k_proj.lora_B.weight torch.Size([1024, 8])
base_model.model.vision_tower.vision_model.encoder.layers.9.self_attn.q_proj.lora_A.weight torch.Size([8, 1024])
base_model.model.vision_tower.vision_model.encoder.layers.9.self_attn.q_proj.lora_B.weight torch.Size([1024, 8])
base_model.model.vision_tower.vision_model.encoder.layers.9.self_attn.v_proj.lora_A.weight torch.Size([8, 1024])
base_model.model.vision_tower.vision_model.encoder.layers.9.self_attn.v_proj.lora_B.weight torch.Size([1024, 8])
```
Thank you again!

**Comment by franciscoliu** [comment]
Sorry about the late reply! I tried to reproduce your error by changing LoRA rank from `8` to `16` but I did not get the same error as you. This may indeed be related to GPU memory constraints as you suggested. You can also try to clear the cache of this previous model `llava-v1.5-hf` and re-run the script.

Regarding the vision encoder being fine-tuned: You've identified a genuine issue. The `find_all_linear_names()` function is looking for specific layer keywords ('gate_proj', 'o_proj', etc.), but it doesn't properly exclude the vision encoder components that might contain those same patterns.

Looking at your safetensors output, it's clear that LoRA adapters are being applied to vision tower components like `vision_tower.vision_model.encoder.layers.9.self_attn.q_proj`. In terms of the safetensors' problem, maybe you can try to modify some of the scripts as below:

1. Explicitly modify the target modules pattern to exclude vision components:
  ```python
  # Add exclusion patterns for vision tower components
  lora_config = LoraConfig(
      target_modules=["q_proj", "v_proj", "k_proj", "o_proj", "gate_proj", "down_proj", "up_proj"],
      modules_to_save=["embed_tokens", "lm_head"],
  )
```
2. Or alternatively, you can try to implement a more selective module finder function that properly filters out vision components:
```python
def find_language_model_linear_names(model):
    """Find all linear layer names in the language model only (excluding vision tower)"""
    linear_names = []
    for name, module in model.named_modules():
        if isinstance(module, nn.Linear) and "vision_tower" not in name:
            if any(keyword in name for keyword in ['gate_proj', 'o_proj', 'up_proj', 'v_proj', 'down_proj', 'k_proj', 'q_proj']):
                linear_names.append(name)
    return linear_names
```
For the empty tensors with higher rank values, maybe you can also try:
- Reducing batch size to accommodate the increased parameters
- Explicitly setting gradient checkpointing if it's not already enabled
- Confirming your VRAM is sufficient for the increased parameter count

**Comment by snowball521** [comment]
Thank you so much qwq. That is really kind of you.
